libs/core/lci_base/src/lci_environment.cpp

```diff
@@ -214,13 +215,21 @@ namespace hpx { namespace util {
             LCI_initialized(&lci_init);
             if (lci_init)
             {
-                HPX_ASSERT(prg_thread_flag.load() == true);
-                HPX_ASSERT(prg_thread_p != nullptr);
+                join_prg_thread_if_running();
+                LCI_finalize();
+            }
+        }
+    }
+
+    void lci_environment::join_prg_thread_if_running()
+    {
+        if (prg_thread_p)
```

Comment:
This change doesn't look right to me. The purpose is to make sure that the thread is joined once. I think the best solution would be to utilize the atomic and write:
```
    bool expected = false;
    if (prg_thread_p.compare_exchange_strong(expected, true))
    {
        prg_thread_p->join();
        prg_thread_p.reset();
    }
```
or even
```
    if (!prg_thread_p.exchange(true))
    {
        prg_thread_p->join();
        prg_thread_p.reset();
    }
```

Reply:
I think they both work. Due to the lock, `prg_thread_p` will only be joined once and then immediately reset to nullptr. After that, no other threads can join the thread again.

Reply:
I think you don't need the lock as you're using that atomic variable.